File: src/gigaloom/ui/schemas/credentials.py
"""Bounded HTTP schemas for the credential operator surface."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


Digest = str
Identity = str


class CredentialScopeResponse(BaseModel):
    audiences: list[str] = Field(max_length=64)
    resources: list[str] = Field(max_length=64)
    operation_classes: list[str] = Field(max_length=64)
    scope_digest: Digest = Field(pattern=r"^[0-9a-f]{64}$")


class CredentialQuotaResponse(BaseModel):
    status: Literal["known", "unknown", "not_applicable"]
    unit: str | None
    limit: int | None = Field(default=None, ge=0)
    remaining: int | None = Field(default=None, ge=0)
    resets_at: str | None
    reason_code: str | None


class CredentialSourceResponse(BaseModel):
    source_id: Identity
    broker_id: Identity
    provider_identity_ref: Identity
    reference_kind: Literal["environment", "keychain", "test"]
    scope: CredentialScopeResponse
    quota: CredentialQuotaResponse
    expires_at: str | None
    state: Literal["current", "expired"]
    demo: bool


class CredentialLeaseResponse(BaseModel):
    lease_id: Identity
    broker_id: Identity
    audience: Identity
    resource: str
    operation_class: Identity
    issued_at: str
    expires_at: str
    parent_run_id: Identity
    scope_digest: Digest = Field(pattern=r"^[0-9a-f]{64}$")
    status: Literal["active", "revoked", "expired"]


class CredentialOperatorSnapshotResponse(BaseModel):
    schema_version: Literal[1] = 1
    revision: Digest = Field(pattern=r"^[0-9a-f]{64}$")
    broker_id: Identity
    sources: list[CredentialSourceResponse] = Field(max_length=64)
    leases: list[CredentialLeaseResponse] = Field(max_length=256)
    content_free: Literal[True]


class CredentialLeaseRequest(BaseModel):
    expected_revision: Digest = Field(pattern=r"^[0-9a-f]{64}$")
    source_id: Identity = Field(min_length=1, max_length=256)
    audience: Identity = Field(min_length=1, max_length=256)
    resource: str = Field(min_length=1, max_length=2_048)
    operation_class: Identity = Field(min_length=1, max_length=256)
    parent_run_id: Identity = Field(min_length=1, max_length=256)
    ttl_seconds: int = Field(ge=1, le=300)


class CredentialRevokeRequest(BaseModel):
    expected_revision: Digest = Field(pattern=r"^[0-9a-f]{64}$")
    confirmed: Literal[True]


class CredentialRevocationReceiptResponse(BaseModel):
    receipt_id: Identity
    lease_id: Identity
    status: Literal["revoked", "already_terminal"]
    reason_code: Identity
    revoked_at: str
    idempotency_key_digest: Digest = Field(pattern=r"^[0-9a-f]{64}$")
    content_free: Literal[True]


class CredentialOperatorActionResponse(BaseModel):
    schema_version: Literal[1] = 1
    state: Literal[
        "admitted",
        "denied",
        "stale",
        "revoked",
        "already_terminal",
    ]
    reason_code: Identity
    snapshot: CredentialOperatorSnapshotResponse
    receipt: CredentialRevocationReceiptResponse | None = None


__all__ = [
    "CredentialLeaseRequest",
    "CredentialOperatorActionResponse",
    "CredentialOperatorSnapshotResponse",
    "CredentialRevokeRequest",
]

File: web/src/api/credentials.ts
import { queryOptions } from "@tanstack/react-query";

import { fetchCockpit, mutateCockpit } from "./core";

export interface CredentialScopeProjection {
  audiences: string[];
  resources: string[];
  operation_classes: string[];
  scope_digest: string;
}

export interface CredentialQuotaProjection {
  status: "known" | "unknown" | "not_applicable";
  unit: string | null;
  limit: number | null;
  remaining: number | null;
  resets_at: string | null;
  reason_code: string | null;
}

export interface CredentialSourceProjection {
  source_id: string;
  broker_id: string;
  provider_identity_ref: string;
  reference_kind: "environment" | "keychain" | "test";
  scope: CredentialScopeProjection;
  quota: CredentialQuotaProjection;
  expires_at: string | null;
  state: "current" | "expired";
  demo: boolean;
}

export interface CredentialLeaseProjection {
  lease_id: string;
  broker_id: string;
  audience: string;
  resource: string;
  operation_class: string;
  issued_at: string;
  expires_at: string;
  parent_run_id: string;
  scope_digest: string;
  status: "active" | "revoked" | "expired";
}

export interface CredentialOperatorSnapshot {
  schema_version: 1;
  revision: string;
  broker_id: string;
  sources: CredentialSourceProjection[];
  leases: CredentialLeaseProjection[];
  content_free: true;
}

export interface CredentialOperatorAction {
  schema_version: 1;
  state: "admitted" | "denied" | "stale" | "revoked" | "already_terminal";
  reason_code: string;
  snapshot: CredentialOperatorSnapshot;
  receipt: {
    receipt_id: string;
    lease_id: string;
    status: "revoked" | "already_terminal";
    reason_code: string;
    revoked_at: string;
    idempotency_key_digest: string;
    content_free: true;
  } | null;
}

const credentialRootKey = ["cockpit", "credential-operator"] as const;

export const credentialKeys = {
  snapshot: () => [...credentialRootKey, "snapshot"] as const,
};

export function credentialSnapshotOptions() {
  return queryOptions({
    queryFn: ({ signal }) => fetchCockpit<CredentialOperatorSnapshot>("/api/credentials", signal),
    queryKey: credentialKeys.snapshot(),
    staleTime: 5_000,
  });
}

export function requestCredentialLease(
  snapshot: CredentialOperatorSnapshot,
  source: CredentialSourceProjection,
): Promise<CredentialOperatorAction> {
  return mutateCockpit<CredentialOperatorAction>("/api/credentials/leases", {
    audience: source.scope.audiences[0],
    expected_revision: snapshot.revision,
    operation_class: source.scope.operation_classes[0],
    parent_run_id: "credential-demo-run",
    resource: source.scope.resources[0],
    source_id: source.source_id,
    ttl_seconds: 300,
  });
}

export function revokeCredentialLease(
  snapshot: CredentialOperatorSnapshot,
  leaseId: string,
): Promise<CredentialOperatorAction> {
  return mutateCockpit<CredentialOperatorAction>(
    `/api/credentials/leases/${encodeURIComponent(leaseId)}/revoke`,
    { confirmed: true, expected_revision: snapshot.revision },
  );
}
